File: 121. Best Time to Buy and Sell Stock/121.cpp
/*
can only consider larger values after current min value
time: O(N)
space: O(1)
*/
class Solution {
public:
    int maxProfit(vector<int>& prices) {
        int minprice = INT_MAX;
        int ans = 0;
        for(int price: prices){
            minprice = min(minprice, price);
            ans = max(ans, price - minprice);
        }
        return ans;
    }
};

File: 121. Best Time to Buy and Sell Stock/121.py
"""
keep track of lowest price seen
time: O(n)
space: O(1)
"""
class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        profit = 0
        min_price = prices[0]

        for price in prices:

            if price < min_price:
                min_price = price
            else:
                profit = max(profit, price - min_price)
        
        return profit
